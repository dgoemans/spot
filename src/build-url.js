export const buildUrl = (base, endpoint, params) => {
  const basis = `${base}/${endpoint}`;
  const queryString = Object.entries(params).reduce((accum, current) => {
    return (
      accum + (accum.length === 0 ? "?" : "&") + `${current[0]}=${current[1]}`
    );
  }, "");
  return `${basis}${queryString}`;
};
