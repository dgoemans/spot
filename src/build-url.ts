export const buildUrl = (base: string, endpoint: string, params = {}) => {
  const basis = `${base}/${endpoint}`;
  const queryString = params
    ? Object.entries(params).reduce((accum, current) => (
      `${accum
          + (accum.length === 0 ? '?' : '&')
      }${current[0]}=${current[1]}`
    ), '')
    : '';
  return `${basis}${queryString}`;
};
