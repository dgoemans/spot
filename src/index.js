import deepmerge from 'deepmerge';
import { makeStore } from './store';

export class Spot {
  constructor(baseUrl, debug) {
    this.store = makeStore(debug);
    this.subscriptions = {};

    this.store.dispatch({
      type: 'SETUP',
      payload: { baseUrl },
    });

    this.store.subscribe(() => {
      if (!this.store.getState().data.loading) {
        Object.values(this.subscriptions).forEach((sub) => {
          sub(this.store.getState());
        });
      }
    });
  }

  query = (
    endpoint,
    params = {},
    path = [endpoint, JSON.stringify(params)],
  ) => {
    this.store.dispatch({
      type: 'QUERY',
      payload: { params, endpoint, path },
    });
  }

  command = (endpoint, params) => {
    this.store.dispatch({
      type: 'COMMAND',
      payload: { params, endpoint },
    });
  }

  subscribeOnce = (subscription) => {
    const hash = subscription.toString();
    this.subscriptions[`${hash}_once`] = (state) => {
      subscription(state);
      delete this.subscriptions[hash];
    };
  }

  subscribe = (subscription) => {
    const hash = subscription.toString();
    this.subscriptions[hash] = subscription;
  }

  unsubscribe = (subscription) => {
    const hash = subscription.toString();
    delete this.subscriptions[hash];
  }

  get = (path) => {
    let state = this.store.getState().data;
    while (path.length) {
      const next = path.shift();
      state = state[next];
    }
    return deepmerge({}, state);
  }

  get data() {
    return deepmerge({}, this.store.getState().data);
  }

  get errors() {
    return deepmerge([], this.store.getState().errors);
  }
}

export const initializeSpot = (baseUrl, debug = false) => new Spot(baseUrl, debug);
