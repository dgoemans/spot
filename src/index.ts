import deepmerge from 'deepmerge';
import { makeStore } from './store';

import { Subscription, ActionConfig, Action } from './types';

/**
 * Subscription callback type
 * @callback SpotSubscriptionCallback
 * @param {any} state
 * @returns {void}
 */

/**
 * Spot instance
 * @type {Spot}
 */
export class Spot {
  store: any;
  subscriptions: { [k: string]: (data: any) => unknown };
  waitForQuery: () => Promise<unknown>;

  constructor(baseUrl: string, debug?: boolean) {
    this.store = makeStore(debug);
    this.subscriptions = {};
    this.waitForQuery = () => new Promise(this.subscribeOnce);

    this.store.dispatch({
      type: 'SETUP',
      payload: { baseUrl },
    });

    this.store.subscribe(() => {
      if (!this.store.getState().data.loading) {
        Object.values(this.subscriptions).forEach((sub) => {
          sub(this.data);
        });
      }
    });
  }

  /**
   * Perform a query, will result in a state update
   * @param {string} endpoint
   * @param {any=} params default: {}
   * @param {string[]=} path default: [endpoint, JSON.stringify(params)]
   */
  query = async (
    endpoint: string,
    params: { [k: string]: unknown } = {},
    path = [endpoint, JSON.stringify(params)],
    config?: ActionConfig,
  ) => {
    this.store.dispatch({
      type: 'QUERY',
      payload: { params, endpoint, path },
      config,
    });
    await this.waitForQuery();
  }

  /**
   * Perform a command, will not result in a state update
   * @param {string} endpoint
   * @param {any=} params default: {}
   */
  command = async (endpoint: string, params: { [k: string]: unknown }, config?: { method?: string }) => {
    this.store.dispatch({
      type: 'COMMAND',
      payload: { params, endpoint },
      config,
    });
    await this.waitForQuery();
  }

  /**
   * Subscribe once to changes
   * @param {SpotSubscriptionCallback} subscription
   */
  subscribeOnce = (subscription: Subscription) => {
    const hash = subscription.toString();
    this.subscriptions[`${hash}_once`] = (state) => {
      subscription(state);
      delete this.subscriptions[hash];
    };
  }

  /**
   * Subscribe to changes
   * @param {SpotSubscriptionCallback} subscription
   */
  subscribe = (subscription: Subscription) => {
    const hash = subscription.toString();
    this.subscriptions[hash] = subscription;
  }

  /**
   * Unsubscribe from changes
   * @param {SpotSubscriptionCallback} subscription
   */
  unsubscribe = (subscription: Subscription) => {
    const hash = subscription.toString();
    delete this.subscriptions[hash];
  }

  /**
   * Get data stored at path
   * @param {string[]} path
   */
  get = (path: string[]) => {
    let state = this.store.getState().data;
    while (path.length) {
      const next = path.shift();
      state = state[next!];
    }
    return deepmerge({}, state);
  }

  /**
   * All the data
   * @readonly
   * @property {any} data
   */
  get data() {
    return deepmerge({}, this.store.getState().data);
  }

  /**
   * List of errors
   * @readonly
   * @property {Error[]} errors
   */
  get errors() {
    return deepmerge([], this.store.getState().errors);
  }
}

/**
 * Initialize a spot instance
 * @param {string} baseUrl
 * @param {boolean} debug default: false
 */
export const initializeSpot = (baseUrl: string, debug: boolean = false) => new Spot(baseUrl, debug);
