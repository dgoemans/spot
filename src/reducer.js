import merge from "deepmerge";

export function commandQuery(state, action) {
  switch (action.type) {
    case "QUERY_COMPLETE":
      return merge(state, action.payload);
    case "COMMAND_COMPLETE":
      return {
        ...state,
      };
    case "QUERY":
    case "COMMAND":
      return {
        ...state,
        loading: true,
      };
    case "STATE_UPDATED":
      return {
        ...state,
        loading: false,
      };
    default:
      return {
        ...state,
      };
  }
}

export function configure(state, action) {
  if (action.type === "SETUP") {
    const { baseUrl } = action.payload;
    return {
      ...state,
      baseUrl,
    };
  }

  return { ...state };
}

function errors(state, action) {
  if (action.type === "ERROR") {
    return [...state, action.payload];
  }
  return [...state];
}

export const reducer = (state, action) => {
  if (state.config && state.config.debug) {
    console.log(
      `Got Action:\n${JSON.stringify(
        action,
        null,
        2
      )}\nCurrent State:\n${JSON.stringify(state, null, 2)}`
    );
  }

  return {
    config: configure(state.config, action),
    data: commandQuery(state.data, action),
    errors: errors(state.errors, action),
  };
};
