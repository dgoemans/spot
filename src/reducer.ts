import merge from 'deepmerge';

import { State, Action } from './types';

export function commandQuery(state: State, action: Action) {
  switch (action.type) {
    case 'QUERY_COMPLETE':
    {
      let current = state;
      // eslint-disable-next-line no-unused-expressions
      action.metadata?.path?.forEach((segment, index) => {
        if (index + 1 < (action.metadata?.path?.length ?? 0)) {
          current = current[segment] || {};
        } else {
          delete current[segment];
        }
      });
      const newState = merge(state, action.payload);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      delete (newState as any).spot.active[`${action.metadata?.correlationId}`];
      return newState;
    }
    case 'COMMAND_COMPLETE':
    {
      const newState = {
        ...state,
      };
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      delete (newState as any).spot.active[`${action.metadata?.correlationId}`];
      return newState;
    }
    case 'QUERY':
      return {
        ...state,
        loading: true,
        spot: {
          ...state.spot,
          active: { ...state.spot.active, [`${action.metadata?.correlationId}`]: true },
        },
      };
    case 'COMMAND':
      return {
        ...state,
        loading: true,
        spot: {
          ...state.spot,
          active: { ...state.spot.active, [`${action.metadata?.correlationId}`]: true },
        },
      };
    case 'ERROR':
    {
      const newState = {
        ...state,
      };
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      delete (newState as any).spot.active[`${action.metadata?.correlationId}`];
      return newState;
    }
    case 'STATE_UPDATED':
      return {
        ...state,
        loading: false, // (Object.keys(state.spot.active).length > 0),
      };
    default:
      return {
        ...state,
      };
  }
}

export function configure(state: State, action: Action) {
  if (action.type === 'SETUP') {
    const { baseUrl } = action.payload;
    return {
      ...state,
      baseUrl,
    };
  }

  return { ...state };
}

function errors(state: State, action: Action) {
  if (action.type === 'ERROR') {
    return [...state, action.payload];
  }
  return [...state];
}

export const reducer = (state: State, action: Action) => {
  if (state.config && state.config.debug) {
    // eslint-disable-next-line no-console
    console.log(
      `Got Action:\n${JSON.stringify(
        action,
        null,
        2,
      )}\nCurrent State:\n${JSON.stringify(state, null, 2)}`,
    );
  }

  return {
    config: configure(state.config, action),
    data: commandQuery(state.data, action),
    errors: errors(state.errors, action),
  };
};
