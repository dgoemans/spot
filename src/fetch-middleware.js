import { buildUrl } from "./build-url";

const fetchJson = async (url, config = { 
  method: 'GET',
  mode: 'cors',
  credentials: 'same-origin',
  headers: {
    'Content-Type': 'application/json'
  }, 
  referrerPolicy: 'no-referrer',
}) => {
  const response = await fetch(url, config);
  return response.json();
};

export const fetchMiddleware = (store) => (next) => async (action) => {
  console.log(`middleware ${JSON.stringify(action, null, 2)}`);
  if (action.type === "QUERY") {
    const url = buildUrl(
      store.getState().config.baseUrl,
      action.payload.endpoint,
      action.payload.params
    );
    try {
      const result = await fetchJson(url);
      store.dispatch({
        type: "QUERY_COMPLETE",
        payload: {
          [action.payload.endpoint]: {
            [btoa(JSON.stringify(action.payload.params))]: {
              ...result,
            },
          },
        },
      });
    } catch (e) {
      console.error(e);
      store.dispatch({ type: "ERROR", payload: e });
    }
  } else if (action.type === "COMMAND") {
    const url = buildUrl(
      store.getState().config.baseUrl,
      action.payload.endpoint,
      action.payload.params
    );
    try {
      const result = await fetchJson(url);
      store.dispatch({
        type: "COMMAND_COMPLETE",
        payload: {
          [action.payload.endpoint]: {
            [action.payload.params]: {
              ...result,
            },
          },
        },
      });
    } catch (e) {
      store.dispatch({ type: "ERROR", payload: e });
    }
  }

  return next(action);
};
