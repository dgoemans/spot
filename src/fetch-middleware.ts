import 'isomorphic-fetch';
import { Dispatch, MiddlewareAPI } from 'redux';

import { Action } from './types';
import { buildUrl } from './build-url';

const defaultFetchConfig: RequestInit = {
  method: 'GET',
  mode: 'cors',
  credentials: 'same-origin',
  headers: {
    'Content-Type': 'application/json',
  },
  referrerPolicy: 'no-referrer',
};

export const fetchMiddleware = (api: MiddlewareAPI) => (next: Dispatch) => async (action: Action) => {
  const nextResult = next(action);

  if (action.type === 'QUERY') {
    const url = buildUrl(
      api.getState().config.baseUrl,
      action.payload.endpoint,
      action.payload.params,
    );

    const correlationId = action.metadata?.correlationId;
    try {
      const response = await fetch(url, {
        ...defaultFetchConfig,
        method: action?.config?.method ?? 'GET',
        headers: {
          ...defaultFetchConfig.headers,
          authorization: action.config?.authorization || '',
        },
      });

      if (response.status < 200 || response.status >= 400) {
        throw new Error(
          `QUERY FAILED ${response.status}: ${response.statusText}`,
        );
      }

      const result = await response.json();

      const payload = {};
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      let depth = payload as any;
      const { path } = action.payload;
      path.forEach((current: string, index: number) => {
        depth[current] = index === path.length - 1 ? result : {};
        depth = depth[current];
      });

      api.dispatch({
        type: 'QUERY_COMPLETE',
        payload,
        metadata: {
          path,
          correlationId,
        },
      });
    } catch (err) {
      // eslint-disable-next-line no-console
      console.error(err);
      api.dispatch({ type: 'ERROR', payload: err.toString(), metadata: { correlationId } });
    } finally {
      api.dispatch({ type: 'STATE_UPDATED', metadata: { correlationId } });
    }
  } else if (action.type === 'COMMAND') {
    const url = buildUrl(
      api.getState().config.baseUrl,
      action.payload.endpoint,
    );

    const correlationId = action.metadata?.correlationId;

    try {
      const response = await fetch(url, {
        ...defaultFetchConfig,
        method: action?.config?.method ?? 'POST',
        body: JSON.stringify(action.payload.params),
        headers: {
          ...defaultFetchConfig.headers,
          authorization: action.config?.authorization || '',
        },
      });

      if (response.status < 200 || response.status >= 400) {
        throw new Error(
          `COMMAND FAILED ${response.status}: ${response.statusText}`,
        );
      }

      api.dispatch({ type: 'COMMAND_COMPLETE', metadata: { correlationId } });
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(e);
      api.dispatch({ type: 'ERROR', payload: e, metadata: { correlationId } });
    } finally {
      api.dispatch({ type: 'STATE_UPDATED', metadata: { correlationId } });
    }
  }

  return nextResult;
};
