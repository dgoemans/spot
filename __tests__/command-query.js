import fetchMock from "jest-fetch-mock";

import { initializeSpot } from "../src";
import { buildUrl } from "../src/build-url";

const users = {
  'id-one': {
    name: "Spot",
    role: "Good Boy",
    age: 7,
  },
  'id-two': {
    name: "Rufus",
    role: "Home Security",
    age: 3,
  }
};

const baseUrl = "http://example.com";

const waitForLoadingDone = (spot) => new Promise(spot.subscribeOnce);

describe("spot", () => {
  beforeEach(() => {
    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockIf(/^http:\/\/example.com.*$/, async (req) => {
      if (req.url.startsWith(`${baseUrl}/fetch-user`)) {
        const tokens = req.url.split('=');
        return {
          status: 200,
          body: JSON.stringify(users[tokens[1]]),
        };
      } else if (req.url.startsWith(`${baseUrl}/update-user`)) {
        const tokens = req.url.split('=').map(token => token.split('&')).reduce((acc, val) => acc.concat(val), []);
        console.log(tokens);
        users[tokens[1]].age = Number.parseInt(tokens[3]);
        return {
          body: "",
          status: 200,
        };
      } else if (req.url.startsWith(`${baseUrl}/invalid-json`)) {
        return {
          body: "<>.aspojdas <XMC98yyfdbshbx hjbgas",
          status: 200,
        };
      } else {
        return {
          status: 404,
          body: "Not Found",
        };
      }
    });
  });

  it("Can fetch data", async () => {
    const spot = initializeSpot(baseUrl);

    {
      spot.query("fetch-user", { userId: 'id-one' }, ['users', 'id-one']);
      expect(spot.getState().data).toStrictEqual({loading: true});
      
      await waitForLoadingDone(spot);

      const expectedResult = {
        "users": { 
            ['id-one']: {"age": 7, "name": "Spot", "role": "Good Boy"}
        },
        loading: false
      };

      expect(spot.getState().data).toStrictEqual(expectedResult);
    }

    {
      spot.query("fetch-user", { userId: 'id-two' }, ['users', 'id-two']);

      await waitForLoadingDone(spot);
      const expectedResult = {
        "users": { 
            'id-one': {"age": 7, "name": "Spot", "role": "Good Boy"},
            'id-two': {"age": 3, "name": "Rufus", "role": "Home Security"}
        },
        loading: false
      };
      expect(spot.getState().data).toStrictEqual(expectedResult);
    }
  });

  it("Adds errors when the fetch fails", async () => {
    const spot = initializeSpot(baseUrl);

    spot.query("invalid-json");
    expect(spot.getState().data).toStrictEqual({loading: true});

    await waitForLoadingDone(spot);

    expect(spot.getState().data).toStrictEqual({loading: false});
    expect(spot.getState().errors).toHaveLength(1);
    expect(spot.getState().errors[0]).toMatchSnapshot();
  });

  it("Can execute commands data", async () => {
    const spot = initializeSpot(baseUrl);

    {
      spot.query("fetch-user", { userId: 'id-two' }, ['users', 'id-two']);

      await waitForLoadingDone(spot);
      const expectedResult = {
        "users": { 
            'id-two': {"age": 3, "name": "Rufus", "role": "Home Security"}
        },
        loading: false
      };
      expect(spot.getState().data).toStrictEqual(expectedResult);
    }

    {
      spot.command("update-user", { userId: "id-two", age: 4 });

      await waitForLoadingDone(spot);
    }

    {
      spot.query("fetch-user", { userId: 'id-two' }, ['users', 'id-two']);

      await waitForLoadingDone(spot);
      const expectedResult = {
        "users": { 
            'id-two': {"age": 4, "name": "Rufus", "role": "Home Security"}
        },
        loading: false
      };
      expect(spot.getState().data).toStrictEqual(expectedResult);
    }
  });
});
