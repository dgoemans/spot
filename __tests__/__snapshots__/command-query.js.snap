// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spot Adds errors when the fetch fails 1`] = `
Object {
  "message": "invalid json response body at  reason: Unexpected token < in JSON at position 0",
  "type": "invalid-json",
}
`;
